django/db/models/base.py

```diff
@@ -609,14 +609,14 @@ def __str__(self):
         return "%s object (%s)" % (self.__class__.__name__, self.pk)
 
     def __eq__(self, other):
+        if self is other:
+            return True
```

Comment:
Comparing instances without primary keys and exactly the same instances is probably the rarest case. I don't think it's desirable to do this every time.

Reply:
Well for items like `sorted`, or a `set`, it will likely match with the same object, so I would not call that the "rarest" case. Especially since it is rather rare that it would work with different models: usually people will call `a == b` in case there is a certain chance that the two are equal.

The idea is that `a is b` might indeed not be very common, but it is a very cheap comparison: normally it will not call `a is b` as a function, but is compiled to a simple reference equality.

Reply:
Model instances without primary keys are not hashable.

Reply:
Indeed, but if these have a primary key, then the `is` check will be sufficient. The point is not that much to move the case of model instances without having a primary key being moved to the top. The point is merely that it covers a larger case that we can now move to the top entirely: the one where the two objects are the same.

Reply:
But as they say in Duch:

> Gissen is missen, meten is weten (Guessing is missing, measuring is knowing)

So I think it would help to simply run the batch of testcases for example a few times with the old and new implementation of `__eq__`. This is likely better than discussions that are all "hyper-hypothetical".

Reply:
They are not hashable so your previous argument about `set`,`dict`, etc. doesn't hold.

> The point is merely that it covers a larger case that we can now move to the top entirely: the one where the two objects are the same.

You want to add check for all `==` comparisons where it will speed up only for the same objects, which is extremely rare (IMO) to compare. You can always start a discussion on the [django-developers mailing list](https://groups.google.com/g/django-developers) (if you don't agree), where you'll reach a wider audience, and see what other think.

Reply:
The point is *not* that they are not hashable, for the very simple reason that `a is b` can also be true if there *is* a primary key.

The `if x is y` check thus does **not** only cover the case when there are no primary keys. It is perfectly possible that these objects *have* primary keys (so these are hashable), and therfore thus is possible.

As said earlier, we did not just "move" the sematics to the top... The `is` check in the rewrite covers more than the original one.